File: python/examples/discovery_demo.py
"""Semantic discovery example for DataGrout Conduit."""

import asyncio
from datagrout.conduit import Client


async def main():
    """Demonstrate semantic discovery."""
    
    async with Client(
        "https://gateway.datagrout.ai/servers/YOUR_SERVER_UUID/mcp",
        auth={"bearer": "YOUR_API_KEY"}
    ) as client:
        
        # Discover tools by query
        print("=== Discovering Tools ===")
        results = await client.discover(
            query="find unpaid invoices",
            limit=10,
            integrations=["salesforce", "quickbooks"]
        )
        
        print(f"Query: {results.query_used}")
        print(f"Found {results.total} tools:\n")
        
        for tool in results.results:
            print(f"  {tool.tool_name}")
            print(f"    Integration: {tool.integration}")
            print(f"    Score: {tool.score:.2f}")
            print(f"    Description: {tool.description}")
            print()
        
        # Perform a tool call
        if results.results:
            tool_name = results.results[0].tool_name
            print(f"=== Executing Top Tool: {tool_name} ===")
            
            result = await client.perform(
                tool=tool_name,
                args={"limit": 5}
            )
            print(f"Result: {result}")
            
            # Check cost
            receipt = client.get_last_receipt()
            if receipt:
                print(f"\nCredits used: {receipt.actual_credits}")


if __name__ == "__main__":
    asyncio.run(main())

File: typescript/examples/discoveryDemo.ts
/**
 * Semantic discovery example for DataGrout Conduit
 */

import { Client } from '@datagrout/conduit';

async function main() {
  const client = new Client({
    url: 'https://gateway.datagrout.ai/servers/YOUR_SERVER_UUID/mcp',
    auth: { bearer: 'YOUR_API_KEY' },
  });

  await client.connect();

  try {
    // Discover tools by query
    console.log('=== Discovering Tools ===');
    const results = await client.discover({
      query: 'find unpaid invoices',
      limit: 10,
      integrations: ['salesforce', 'quickbooks'],
    });

    console.log(`Query: ${results.queryUsed}`);
    console.log(`Found ${results.total} tools:\n`);

    for (const tool of results.results) {
      console.log(`  ${tool.toolName}`);
      console.log(`    Integration: ${tool.integration}`);
      console.log(`    Score: ${tool.score?.toFixed(2)}`);
      console.log(`    Description: ${tool.description}`);
      console.log();
    }

    // Perform a tool call
    if (results.results.length > 0) {
      const toolName = results.results[0].toolName;
      console.log(`=== Executing Top Tool: ${toolName} ===`);

      const result = await client.perform({
        tool: toolName,
        args: { limit: 5 },
      });
      console.log('Result:', result);

      // Check cost
      const receipt = client.getLastReceipt();
      if (receipt) {
        console.log(`\nCredits used: ${receipt.actualCredits}`);
      }
    }
  } finally {
    await client.disconnect();
  }
}

main().catch(console.error);
